File: Platform/Workspace/Typescript/src/allors/framework/database/data/Instanceof.ts
import { ObjectType, PropertyType, ObjectTypeRef } from '../../meta';

import { Predicate } from './Predicate';

export class Instanceof implements Predicate {
  public propertyType: PropertyType;
  public objectType: ObjectType | ObjectTypeRef;

  constructor(fields?: Partial<Instanceof> | PropertyType) {
    if ((fields as PropertyType).objectType) {
      this.propertyType = fields as PropertyType;
    } else {
      Object.assign(this, fields);
    }
  }

  public toJSON(): any {

    const metaObjectType = this.objectType as ObjectTypeRef;
    const objectType = metaObjectType.objectType ? metaObjectType.objectType : this.objectType as ObjectType;

    return {
      kind: 'Instanceof',
      propertytype: this.propertyType.id,
      objecttype: objectType.id,
    };
  }
}

File: Core/Workspace/CSharp/Domain/Export/Core/Data/Instanceof.cs
//-------------------------------------------------------------------------------------------------
// <copyright file="Instanceof.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
//-------------------------------------------------------------------------------------------------

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;
    using Allors.Protocol.Data;
    using Allors.Workspace.Meta;

    public class Instanceof : IPropertyPredicate
    {
        public string Parameter { get; set; }

        public Instanceof(IComposite objectType = null) => this.ObjectType = objectType;

        public IComposite ObjectType { get; set; }

        public IPropertyType PropertyType { get; set; }

        public Predicate ToJson() =>
            new Predicate
            {
                Kind = PredicateKind.Instanceof,
                ObjectType = this.ObjectType?.Id,
                PropertyType = this.PropertyType?.Id,
            };

        bool IPredicate.ShouldTreeShake(IReadOnlyDictionary<string, object> arguments) => ((IPredicate)this).HasMissingArguments(arguments);

        bool IPredicate.HasMissingArguments(IReadOnlyDictionary<string, object> arguments) => this.Parameter != null && (arguments == null || !arguments.ContainsKey(this.Parameter));
    }
}
